lib/code_corps/github/event/installation/repos.ex

```diff
@@ -7,76 +7,155 @@ defmodule CodeCorps.GitHub.Event.Installation.Repos do
     GithubAppInstallation,
     GithubRepo,
     GitHub.Installation,
+    GitHub,
     Repo
   }
 
   alias CodeCorps.GitHub.Adapters.GithubRepo, as: GithubRepoAdapter
 
   alias Ecto.{Changeset, Multi}
 
-  @typep outcome :: {:ok, list(GithubRepo.t)} |
-                    {:error, CodeCorps.GitHub.api_error_struct} |
-                    {:error, :invalid_repo_payload}
+  @typep repo_result :: {:ok | :error, GithubRepo.t | Changeset.t}
+  @typep aggregated_result :: {:ok | :error, list(GithubRepo.t | Changeset.t)}
+
+  @doc ~S"""
+  Marks a `GithubAppInstallation` as "processing" and immediately returns it.
+
+  In the background, it fires a task to asynchronously call &process/1
+  """
+  @spec process_async(GithubAppInstallation.t) :: {:ok, GithubAppInstallation.t, Task.t}
+  def process_async(%GithubAppInstallation{} = installation) do
+    {:ok, %GithubAppInstallation{} = processing_installation} = installation |> set_state("processing")
+
+    task = Task.Supervisor.async(:background_processor, fn -> processing_installation |> process() end)
+
+    {:ok, processing_installation, task}
+  end
 
   @doc ~S"""
   Fetches a list of repositories for a `GithubAppInstallation` from the GitHub
-  API, then creates a `GithubRepo` record for each of those repositories.
+  API and matches up `GithubRepo` records stored locally using that list as the
+  master list.
   """
-  @spec process(GithubAppInstallation.t) :: outcome
+  @spec process(GithubAppInstallation.t) :: {:ok | :error, GithubAppInstallation.t}
   def process(%GithubAppInstallation{} = installation) do
     multi =
       Multi.new
-      |> Multi.update(:processing_installation, installation |> Changeset.change(%{state: "processing"}))
-      |> Multi.run(:process_repos, &process_repos/1)
-      |> Multi.run(:stop_processing, &stop_processing/1)
+      |> Multi.run(:processing_installation, fn _ -> {:ok, installation} end)
+      |> Multi.run(:api_response, &fetch_api_repo_list/1)
+      |> Multi.run(:repo_attrs_list, &adapt_api_repo_list/1)
+      |> Multi.run(:deleted_repos, &delete_repos/1)
+      |> Multi.run(:synced_repos, &sync_repos/1)
+      |> Multi.run(:processed_installation, &mark_processed/1)
```

Comment:
Instead of doing a few large steps, which wrap several minor steps, I tried to do this as explicitly as I can.